File: GLLara/GLLASCIIScanner.m
//
//  GLLASCIIScanner.m
//  GLLara
//
//  Created by Torsten Kammer on 01.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLASCIIScanner.h"

@interface GLLASCIIScanner ()
{
	NSScanner *scanner;
}

- (NSInteger)_readInteger;
- (void)_skipComments;

@property (nonatomic, readwrite) BOOL isValid;

@end

@implementation GLLASCIIScanner

- (id)initWithString:(NSString *)string;
{
	if (!(self = [super init])) return nil;
	
	scanner = [NSScanner scannerWithString:string];
	
	// Use american english at all times, because that is the number format used.
	scanner.locale = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US"];
	self.isValid = YES;
	
	return self;
}

- (uint32_t)readUint32;
{
	return (uint32_t) [self _readInteger];
}
- (uint16_t)readUint16;
{
	return (uint16_t) [self _readInteger];
}
- (uint8_t)readUint8;
{
	return (uint8_t) [self _readInteger];
}
- (Float32)readFloat32;
{
	[self _skipComments];
	if (scanner.isAtEnd)
	{
		self.isValid = NO;
		return 0.0f;
	}
	
	float result;
	if (![scanner scanFloat:&result])
	{
		self.isValid = NO;
		return 0.0f;
	}
	
	return result;
}

- (NSString *)readPascalString;
{
	[self _skipComments];
	if (scanner.isAtEnd)
	{
		self.isValid = NO;
		return nil;
	}
	
	[scanner scanCharactersFromSet:[NSCharacterSet whitespaceAndNewlineCharacterSet] intoString:NULL];
	
	NSString *result;
	if (![scanner scanUpToCharactersFromSet:[NSCharacterSet newlineCharacterSet] intoString:&result])
	{
		self.isValid = NO;
		return nil;
	}
	
	return result;
}


- (NSInteger)_readInteger;
{
	[self _skipComments];
	if (scanner.isAtEnd)
	{
		self.isValid = NO;
		return 0;
	}
	
	NSInteger result;
	if (![scanner scanInteger:&result])
		[NSException raise:NSInvalidArgumentException format:@"Expected integer at position %lu, but didn't get it.", scanner.scanLocation];
	
	return result;
}

- (void)_skipComments;
{
	while ([scanner scanString:@"#" intoString:NULL])
		[scanner scanUpToCharactersFromSet:[NSCharacterSet newlineCharacterSet] intoString:NULL];
}

@end

File: GLLara/GLLASCIIScanner.swift
//
//  GLLASCIIScanner.m
//  GLLara
//
//  Created by Torsten Kammer on 01.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//
import Foundation

/*!
 * @abstract Reader for the .mesh.ascii format.
 * @discussion It was deliberately designed to have the same interface as the
 * TRInDataStream. Thus, it can read integers in different widths, although
 * parsing them from an ASCII file is always the same work.
 */
class GLLASCIIScanner: GLLDataReader {
    init(string: String) {
        scanner = Scanner(string: string)
        // Use american english at all times, because that is the number format used.
        scanner.locale = Locale(identifier: "en_US")
    }
    
    private let scanner: Scanner
    
    func readUint32() -> UInt32 {
        return UInt32(readInteger())
    }
    
    func readUint16() -> UInt16 {
        return UInt16(readInteger())
    }
    
    func readInt16() -> Int16 {
        return Int16(readInteger())
    }
    
    func readUint8() -> UInt8 {
        return UInt8(readInteger())
    }
    
    func readFloat32() -> Float32 {
        skipComments()
        if scanner.isAtEnd {
            isValid = false
            return 0.0
        }
        
        guard let result = scanner.scanFloat() else {
            if (scanner.scanString("NaN") != nil) {
                // Haha, very funny. Idiots.
                return Float.nan
            } else {
                isValid = false
                return 0.0
            }
        }
        return result
    }
    
    func readPascalString() -> String {
        skipComments()
        if scanner.isAtEnd {
            isValid = false
            return ""
        }
        
        _ = scanner.scanCharacters(from: CharacterSet.whitespacesAndNewlines)
        guard let result = scanner.scanUpToCharacters(from: CharacterSet.newlines) else {
            isValid = false
            return ""
        }
        return result
    }
    
    func hasNewline() -> Bool {
        // Skip only whitespace, not newline, because the scanner won't recognize the newline otherwise
        scanner.charactersToBeSkipped = CharacterSet.whitespaces
        if scanner.scanString("#") != nil {
            // Has a comment, which ends a line.
            _ = scanner.scanUpToCharacters(from: CharacterSet.newlines)
            scanner.charactersToBeSkipped = CharacterSet.whitespacesAndNewlines
            return true
        } else if scanner.scanCharacters(from: CharacterSet.newlines) != nil {
            // Has newline, which obviously ends a line.
            scanner.charactersToBeSkipped = CharacterSet.whitespacesAndNewlines
            return true
        }
        scanner.charactersToBeSkipped = CharacterSet.whitespacesAndNewlines
        return false
    }
    
    var isValid: Bool = true
    
    private func skipComments() {
        while scanner.scanString("#") != nil {
            _ = scanner.scanUpToCharacters(from: CharacterSet.newlines)
        }
    }
    
    private func readInteger() -> Int {
        skipComments()
        if scanner.isAtEnd {
            isValid = false
            return 0
        }
        
        guard let result = scanner.scanInt() else {
            isValid = false
            return 0
        }
        return result
    }
}
